File: src/bindings/python/hammer_tests.py
import unittest
import hammer as h

class TestTokenParser(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.token("95\xa2")
    def test_success(self):
        self.assertEqual(self.parser.parse("95\xa2"), "95\xa2")
    def test_partial_fails(self):
        self.assertEqual(self.parser.parse("95"), None)

class TestChParser(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser_int = h.ch(0xa2)
        cls.parser_chr = h.ch("\xa2")
    def test_success(self):
        self.assertEqual(self.parser_int.parse("\xa2"), 0xa2)
        self.assertEqual(self.parser_chr.parse("\xa2"), "\xa2")
    def test_failure(self):
        self.assertEqual(self.parser_int.parse("\xa3"), None)
        self.assertEqual(self.parser_chr.parse("\xa3"), None)

class TestChRange(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.ch_range("a", "c")
    def test_success(self):
        self.assertEqual(self.parser.parse("b"), "b")
    def test_failure(self):
        self.assertEqual(self.parser.parse("d"), None)

class TestInt64(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.int64()
    def test_success(self):
        self.assertEqual(self.parser.parse("\xff\xff\xff\xfe\x00\x00\x00\x00"), -0x200000000)
    def test_failure(self):
        self.assertEqual(self.parser.parse("\xff\xff\xff\xfe\x00\x00\x00"), None)

class TestInt32(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.int32()
    def test_success(self):
        self.assertEqual(self.parser.parse("\xff\xfe\x00\x00"), -0x20000)
        self.assertEqual(self.parser.parse("\x00\x02\x00\x00"), 0x20000)
    def test_failure(self):
        self.assertEqual(self.parser.parse("\xff\xfe\x00"), None)
        self.assertEqual(self.parser.parse("\x00\x02\x00"), None)

class TestInt16(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.int16()
    def test_success(self):
        self.assertEqual(self.parser.parse("\xfe\x00"), -0x200)
        self.assertEqual(self.parser.parse("\x02\x00"), 0x200)
    def test_failure(self):
        self.assertEqual(self.parser.parse("\xfe"), None)
        self.assertEqual(self.parser.parse("\x02"), None)

class TestInt8(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.int8()
    def test_success(self):
        self.assertEqual(self.parser.parse("\x88"), -0x78)
    def test_failure(self):
        self.assertEqual(self.parser.parse(""), None)

class TestUint64(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.uint64()
    def test_success(self):
        self.assertEqual(self.parser.parse("\x00\x00\x00\x02\x00\x00\x00\x00"), 0x200000000)
    def test_failure(self):
        self.assertEqual(self.parser.parse("\x00\x00\x00\x02\x00\x00\x00"), None)

class TestUint32(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.uint32()
    def test_success(self):
        self.assertEqual(self.parser.parse("\x00\x02\x00\x00"), 0x20000)
    def test_failure(self):
        self.assertEqual(self.parser.parse("\x00\x02\x00"), None)

class TestUint16(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.uint16()
    def test_success(self):
        self.assertEqual(self.parser.parse("\x02\x00"), 0x200)
    def test_failure(self):
        self.assertEqual(self.parser.parse("\x02"), None)

class TestUint8(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.uint8()
    def test_success(self):
        self.assertEqual(self.parser.parse("\x78"), 0x78)
    def test_failure(self):
        self.assertEqual(self.parser.parse(""), None)
        
class TestIntRange(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.int_range(h.uint8(), 3, 10)
    def test_success(self):
        self.assertEqual(self.parser.parse("\x05"), 5)
    def test_failure(self):
        self.assertEqual(self.parser.parse("\x0b"), None)

class TestWhitespace(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.whitespace(h.ch("a"))
    def test_success(self):
        self.assertEqual(self.parser.parse("a"), "a")
        self.assertEqual(self.parser.parse(" a"), "a")
        self.assertEqual(self.parser.parse("  a"), "a")
        self.assertEqual(self.parser.parse("\ta"), "a")
    def test_failure(self):
        self.assertEqual(self.parser.parse("_a"), None)

class TestWhitespaceEnd(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.whitespace(h.end_p())
    def test_success(self):
        self.assertEqual(self.parser.parse(""), None) # empty string
        self.assertEqual(self.parser.parse("  "), None) # empty string
    def test_failure(self):
        self.assertEqual(self.parser.parse("  x"), None)

class TestLeft(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.left(h.ch("a"), h.ch(" "))
    def test_success(self):
        self.assertEqual(self.parser.parse("a "), "a")
    def test_failure(self):
        self.assertEqual(self.parser.parse("a"), None)
        self.assertEqual(self.parser.parse(" "), None)
        self.assertEqual(self.parser.parse("ab"), None)

class TestRight(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.right(h.ch(" "), h.ch("a"))
    def test_success(self):
        self.assertEqual(self.parser.parse(" a"), "a")
    def test_failure(self):
        self.assertEqual(self.parser.parse("a"), None)
        self.assertEqual(self.parser.parse(" "), None)
        self.assertEqual(self.parser.parse("ba"), None)

class TestMiddle(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.middle(h.ch(" "), h.ch("a"), h.ch(" "))
    def test_success(self):
        self.assertEqual(self.parser.parse(" a "), "a")
    def test_failure(self):
        self.assertEqual(self.parser.parse("a"), None)
        self.assertEqual(self.parser.parse(" "), None)
        self.assertEqual(self.parser.parse(" a"), None)
        self.assertEqual(self.parser.parse("a "), None)
        self.assertEqual(self.parser.parse(" b "), None)
        self.assertEqual(self.parser.parse("ba "), None)
        self.assertEqual(self.parser.parse(" ab"), None)

class TestAction(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.action(h.sequence(h.choice(h.ch("a"), h.ch("A")),
                                         h.choice(h.ch("b"), h.ch("B"))),
                                lambda x: [y.upper() for y in x])
    def test_success(self):
        self.assertEqual(self.parser.parse("ab"), ["A", "B"])
        self.assertEqual(self.parser.parse("AB"), ["A", "B"])
    def test_failure(self):
        self.assertEqual(self.parser.parse("XX"), None)

class TestIn(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.in_("abc")
    def test_success(self):
        self.assertEqual(self.parser.parse("b"), "b") 
    def test_failure(self):
        self.assertEqual(self.parser.parse("d"), None)

class TestNotIn(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.not_in("abc")
    def test_success(self):
        self.assertEqual(self.parser.parse("d"), "d")
    def test_failure(self):
        self.assertEqual(self.parser.parse("a"), None)

class TestEndP(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.sequence(h.ch("a"), h.end_p())
    def test_success(self):
        self.assertEqual(self.parser.parse("a"), ("a",))
    def test_failure(self):
        self.assertEqual(self.parser.parse("aa"), None)

class TestNothingP(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.nothing_p()
    def test_success(self):
        pass
    def test_failure(self):
        self.assertEqual(self.parser.parse("a"), None)

class TestSequence(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.sequence(h.ch("a"), h.ch("b"))
    def test_success(self):
        self.assertEqual(self.parser.parse("ab"), ('a','b'))
    def test_failure(self):
        self.assertEqual(self.parser.parse("a"), None)
        self.assertEqual(self.parser.parse("b"), None)

class TestSequenceWhitespace(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.sequence(h.ch("a"), h.whitespace(h.ch("b")))
    def test_success(self):
        self.assertEqual(self.parser.parse("ab"), ('a','b'))
        self.assertEqual(self.parser.parse("a b"), ('a','b'))
        self.assertEqual(self.parser.parse("a  b"), ('a','b'))
    def test_failure(self):
        self.assertEqual(self.parser.parse("a  c"), None)

class TestChoice(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.choice(h.ch("a"), h.ch("b"))
    def test_success(self):
        self.assertEqual(self.parser.parse("a"), "a")
        self.assertEqual(self.parser.parse("b"), "b")
    def test_failure(self):
        self.assertEqual(self.parser.parse("c"), None)

class TestButNot(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.butnot(h.ch("a"), h.token("ab"))
    def test_success(self):
        self.assertEqual(self.parser.parse("a"), "a")
        self.assertEqual(self.parser.parse("aa"), "a")
    def test_failure(self):
        self.assertEqual(self.parser.parse("ab"), None)

class TestButNotRange(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.butnot(h.ch_range("0", "9"), h.ch("6"))
    def test_success(self):
        self.assertEqual(self.parser.parse("4"), "4")
    def test_failure(self):
        self.assertEqual(self.parser.parse("6"), None)

class TestDifference(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.difference(h.token("ab"), h.ch("a"))
    def test_success(self):
        self.assertEqual(self.parser.parse("ab"), "ab")
    def test_failure(self):
        self.assertEqual(self.parser.parse("a"), None)

class TestXor(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.xor(h.ch_range("0", "6"), h.ch_range("5", "9"))
    def test_success(self):
        self.assertEqual(self.parser.parse("0"), "0")
        self.assertEqual(self.parser.parse("9"), "9")
    def test_failure(self):
        self.assertEqual(self.parser.parse("5"), None)
        self.assertEqual(self.parser.parse("a"), None)

class TestMany(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.many(h.choice(h.ch("a"), h.ch("b")))
    def test_success(self):
        self.assertEqual(self.parser.parse(""), ())
        self.assertEqual(self.parser.parse("a"), ('a',))
        self.assertEqual(self.parser.parse("b"), ('b',))
        self.assertEqual(self.parser.parse("aabbaba"), ('a','a','b','b','a','b','a'))
    def test_failure(self):
        pass

class TestMany1(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.many1(h.choice(h.ch("a"), h.ch("b")))
    def test_success(self):
        self.assertEqual(self.parser.parse("a"), ("a",))
        self.assertEqual(self.parser.parse("b"), ("b",))
        self.assertEqual(self.parser.parse("aabbaba"), ("a", "a", "b", "b", "a", "b", "a"))
    def test_failure(self):
        self.assertEqual(self.parser.parse(""), None)
        self.assertEqual(self.parser.parse("daabbabadef"), None)

class TestRepeatN(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.repeat_n(h.choice(h.ch("a"), h.ch("b")), 2)
    def test_success(self):
        self.assertEqual(self.parser.parse("abdef"), ('a', 'b'))
    def test_failure(self):
        self.assertEqual(self.parser.parse("adef"), None)
        self.assertEqual(self.parser.parse("dabdef"), None)

class TestOptional(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.sequence(h.ch("a"), h.optional(h.choice(h.ch("b"), h.ch("c"))), h.ch("d"))
    def test_success(self):
        self.assertEqual(self.parser.parse("abd"), ('a','b','d'))
        self.assertEqual(self.parser.parse("acd"), ('a','c','d'))
        self.assertEqual(self.parser.parse("ad"), ('a',h.Placeholder(), 'd'))
    def test_failure(self):
        self.assertEqual(self.parser.parse("aed"), None)
        self.assertEqual(self.parser.parse("ab"), None)
        self.assertEqual(self.parser.parse("ac"), None)

class TestIgnore(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.sequence(h.ch("a"), h.ignore(h.ch("b")), h.ch("c"))
    def test_success(self):
        self.assertEqual(self.parser.parse("abc"), ("a","c"))
    def test_failure(self):
        self.assertEqual(self.parser.parse("ac"), None)

class TestSepBy(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.sepBy(h.choice(h.ch("1"), h.ch("2"), h.ch("3")), h.ch(","))
    def test_success(self):
        self.assertEqual(self.parser.parse("1,2,3"), ('1','2','3'))
        self.assertEqual(self.parser.parse("1,3,2"), ('1','3','2'))
        self.assertEqual(self.parser.parse("1,3"), ('1','3'))
        self.assertEqual(self.parser.parse("3"), ('3',))
        self.assertEqual(self.parser.parse(""), ())
    def test_failure(self):
        pass

class TestSepBy1(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.sepBy1(h.choice(h.ch("1"), h.ch("2"), h.ch("3")), h.ch(","))
    def test_success(self):
        self.assertEqual(self.parser.parse("1,2,3"), ('1','2','3'))
        self.assertEqual(self.parser.parse("1,3,2"), ('1','3','2'))
        self.assertEqual(self.parser.parse("1,3"), ('1','3'))
        self.assertEqual(self.parser.parse("3"), ('3',))
    def test_failure(self):
        self.assertEqual(self.parser.parse(""), None)

class TestEpsilonP1(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.sequence(h.ch("a"), h.epsilon_p(), h.ch("b"))
    def test_success(self):
        self.assertEqual(self.parser.parse("ab"), ("a", "b"))
    def test_failure(self):
        pass

class TestEpsilonP2(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.sequence(h.epsilon_p(), h.ch("a"))
    def test_success(self):
        self.assertEqual(self.parser.parse("a"), ("a",))
    def test_failure(self):
        pass

class TestEpsilonP3(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.sequence(h.ch("a"), h.epsilon_p())
    def test_success(self):
        self.assertEqual(self.parser.parse("a"), ("a",))
    def test_failure(self):
        pass

class TestAttrBool(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.attr_bool(h.many1(h.choice(h.ch("a"), h.ch("b"))),
                                 lambda x: x[0] == x[1])
    def test_success(self):
        self.assertEqual(self.parser.parse("aa"), ("a", "a"))
        self.assertEqual(self.parser.parse("bb"), ("b", "b"))
    def test_failure(self):
        self.assertEqual(self.parser.parse("ab"), None)

class TestAnd1(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.sequence(h.and_(h.ch("0")), h.ch("0"))
    def test_success(self):
        self.assertEqual(self.parser.parse("0"), ("0",))
    def test_failure(self):
        pass

class TestAnd2(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.sequence(h.and_(h.ch("0")), h.ch("1"))
    def test_success(self):
        pass
    def test_failure(self):
        self.assertEqual(self.parser.parse("0"), None)

class TestAnd3(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.sequence(h.ch("1"), h.and_(h.ch("2")))
    def test_success(self):
        self.assertEqual(self.parser.parse("12"), ('1',))
    def test_failure(self):
        pass

class TestNot1(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.sequence(h.ch("a"),
                                h.choice(h.ch("+"), h.token("++")),
                                h.ch("b"))
    def test_success(self):
        self.assertEqual(self.parser.parse("a+b"), ("a", "+", "b"))
    def test_failure(self):
        self.assertEqual(self.parser.parse("a++b"), None)

class TestNot2(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.parser = h.sequence(h.ch("a"), h.choice(h.sequence(h.ch("+"), h.not_(h.ch("+"))),
                                                    h.token("++")),
                                h.ch("b"))
    def test_success(self):
        self.assertEqual(self.parser.parse("a+b"), ('a', ('+',), 'b'))
        self.assertEqual(self.parser.parse("a++b"), ('a', "++", 'b'))
    def test_failure(self):
        pass

# ### this is commented out for packrat in C ...
# #class TestLeftrec(unittest.TestCase):
# #    @classmethod
# #    def setUpClass(cls):
# #        cls.parser = h.indirect()
# #        a = h.ch("a")
# #        h.bind_indirect(cls.parser, h.choice(h.sequence(cls.parser, a), a))
# #    def test_success(self):
# #        self.assertEqual(self.parser.parse("a"), "a")
# #        self.assertEqual(self.parser.parse("aa"), ["a", "a"])
# #        self.assertEqual(self.parser.parse("aaa"), ["a", "a", "a"])
# #    def test_failure(self):
# #        pass


class TestRightrec(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        #raise unittest.SkipTest("Bind doesn't work right now")
        cls.parser = h.indirect()
        a = h.ch("a")
        cls.parser.bind(h.choice(h.sequence(a, cls.parser),
                                 h.epsilon_p()))
    def test_success(self):
        self.assertEqual(self.parser.parse("a"), ('a',))
        self.assertEqual(self.parser.parse("aa"), ('a', ('a',)))
        self.assertEqual(self.parser.parse("aaa"), ('a', ('a', ('a',))))
    def test_failure(self):
        pass

# ### this is just for GLR
# #class TestAmbiguous(unittest.TestCase):
# #    @classmethod
# #    def setUpClass(cls):
# #        cls.parser = h.indirect()
# #        d = h.ch("d")
# #        p = h.ch("+")
# #        h.bind_indirect(cls.parser, h.choice(h.sequence(cls.parser, p, cls.parser), d))
# #        # this is supposed to be flattened
# #    def test_success(self):
# #        self.assertEqual(self.parser.parse("d"), ["d"])
# #        self.assertEqual(self.parser.parse("d+d"), ["d", "+", "d"])
# #        self.assertEqual(self.parser.parse("d+d+d"), ["d", "+", "d", "+", "d"])
# #    def test_failure(self):
# #        self.assertEqual(self.parser.parse("d+"), None)

File: src/bindings/dotnet/test/hammer_tests.cs
namespace Hammer.Test {
    using NUnit.Framework;
    [TestFixture]
    public partial class HammerTest {
        [Test]
        public void TestToken() {
            Parser parser;
            parser = Hammer.Token("95\xa2");
              CheckParseOK(parser, "95\xa2", new byte[]{ 0x39, 0x35, 0xa2});
              CheckParseFail(parser, "95\xa2");
        }
        [Test]
        public void TestCh() {
            Parser parser;
            parser = Hammer.Ch(0xa2);
              CheckParseOK(parser, "\xa2", '\xa2');
              CheckParseFail(parser, "\xa3");
        }
        [Test]
        public void TestCh_range() {
            Parser parser;
            parser = Hammer.Ch_range(0x61, 0x63);
              CheckParseOK(parser, "b", 'b');
              CheckParseFail(parser, "d");
        }
        [Test]
        public void TestInt64() {
            Parser parser;
            parser = Hammer.Int64();
              CheckParseOK(parser, "\xff\xff\xff\xfe\x00\x00\x00\x00", (System.Int64)(-0x200000000));
              CheckParseFail(parser, "\xff\xff\xff\xfe\x00\x00\x00");
        }
        [Test]
        public void TestInt32() {
            Parser parser;
            parser = Hammer.Int32();
              CheckParseOK(parser, "\xff\xfe\x00\x00", (System.Int64)(-0x20000));
              CheckParseFail(parser, "\xff\xfe\x00");
              CheckParseOK(parser, "\x00\x02\x00\x00", (System.Int64)(0x20000));
              CheckParseFail(parser, "\x00\x02\x00");
        }
        [Test]
        public void TestInt16() {
            Parser parser;
            parser = Hammer.Int16();
              CheckParseOK(parser, "\xfe\x00", (System.Int64)(-0x200));
              CheckParseFail(parser, "\xfe");
              CheckParseOK(parser, "\x02\x00", (System.Int64)(0x200));
              CheckParseFail(parser, "\x02");
        }
        [Test]
        public void TestInt8() {
            Parser parser;
            parser = Hammer.Int8();
              CheckParseOK(parser, "\x88", (System.Int64)(-0x78));
              CheckParseFail(parser, "");
        }
        [Test]
        public void TestUint64() {
            Parser parser;
            parser = Hammer.Uint64();
              CheckParseOK(parser, "\x00\x00\x00\x02\x00\x00\x00\x00", (System.UInt64)0x200000000);
              CheckParseFail(parser, "\x00\x00\x00\x02\x00\x00\x00");
        }
        [Test]
        public void TestUint32() {
            Parser parser;
            parser = Hammer.Uint32();
              CheckParseOK(parser, "\x00\x02\x00\x00", (System.UInt64)0x20000);
              CheckParseFail(parser, "\x00\x02\x00");
        }
        [Test]
        public void TestUint16() {
            Parser parser;
            parser = Hammer.Uint16();
              CheckParseOK(parser, "\x02\x00", (System.UInt64)0x200);
              CheckParseFail(parser, "\x02");
        }
        [Test]
        public void TestUint8() {
            Parser parser;
            parser = Hammer.Uint8();
              CheckParseOK(parser, "x", (System.UInt64)0x78);
              CheckParseFail(parser, "");
        }
        [Test]
        public void TestInt_range() {
            Parser parser;
            parser = Hammer.Int_range(Hammer.Uint8(), 0x3, 0x10);
              CheckParseOK(parser, "\x05", (System.UInt64)0x5);
              CheckParseFail(parser, "\x0b");
        }
        [Test]
        public void TestWhitespace() {
            Parser parser;
            parser = Hammer.Whitespace(Hammer.Ch(0x61));
              CheckParseOK(parser, "a", 'a');
              CheckParseOK(parser, " a", 'a');
              CheckParseOK(parser, "  a", 'a');
              CheckParseOK(parser, "\x09a", 'a');
              CheckParseFail(parser, "_a");
            parser = Hammer.Whitespace(Hammer.End_p());
              CheckParseOK(parser, "", null);
              CheckParseOK(parser, "  ", null);
              CheckParseFail(parser, "  x");
        }
        [Test]
        public void TestLeft() {
            Parser parser;
            parser = Hammer.Left(Hammer.Ch(0x61), Hammer.Ch(0x20));
              CheckParseOK(parser, "a ", 'a');
              CheckParseFail(parser, "a");
              CheckParseFail(parser, " ");
              CheckParseFail(parser, "ba");
        }
        [Test]
        public void TestMiddle() {
            Parser parser;
            parser = Hammer.Middle(Hammer.Ch(' '), Hammer.Ch('a'), Hammer.Ch(' '));
              CheckParseOK(parser, " a ", 'a');
              CheckParseFail(parser, "a");
              CheckParseFail(parser, " a");
              CheckParseFail(parser, "a ");
              CheckParseFail(parser, " b ");
              CheckParseFail(parser, "ba ");
              CheckParseFail(parser, " ab");
        }
        [Test]
        public void TestIn() {
            Parser parser;
            parser = Hammer.In("abc");
              CheckParseOK(parser, "b", 'b');
              CheckParseFail(parser, "d");
        }
        [Test]
        public void TestNot_in() {
            Parser parser;
            parser = Hammer.Not_in("abc");
              CheckParseOK(parser, "d", 'd');
              CheckParseFail(parser, "a");
        }
        [Test]
        public void TestEnd_p() {
            Parser parser;
            parser = Hammer.Sequence(Hammer.Ch('a'), Hammer.End_p());
              CheckParseOK(parser, "a", new object[]{ 'a'});
              CheckParseFail(parser, "aa");
        }
        [Test]
        public void TestNothing_p() {
            Parser parser;
            parser = Hammer.Nothing_p();
              CheckParseFail(parser, "a");
        }
        [Test]
        public void TestSequence() {
            Parser parser;
            parser = Hammer.Sequence(Hammer.Ch('a'), Hammer.Ch('b'));
              CheckParseOK(parser, "ab", new object[]{ 'a', 'b'});
              CheckParseFail(parser, "a");
              CheckParseFail(parser, "b");
            parser = Hammer.Sequence(Hammer.Ch('a'), Hammer.Whitespace(Hammer.Ch('b')));
              CheckParseOK(parser, "ab", new object[]{ 'a', 'b'});
              CheckParseOK(parser, "a b", new object[]{ 'a', 'b'});
              CheckParseOK(parser, "a  b", new object[]{ 'a', 'b'});
        }
        [Test]
        public void TestChoice() {
            Parser parser;
            parser = Hammer.Choice(Hammer.Ch('a'), Hammer.Ch('b'));
              CheckParseOK(parser, "a", 'a');
              CheckParseOK(parser, "b", 'b');
              CheckParseOK(parser, "ab", 'a');
              CheckParseFail(parser, "c");
        }
        [Test]
        public void TestButnot() {
            Parser parser;
            parser = Hammer.Butnot(Hammer.Ch('a'), Hammer.Token("ab"));
              CheckParseOK(parser, "a", 'a');
              CheckParseFail(parser, "ab");
              CheckParseOK(parser, "aa", 'a');
            parser = Hammer.Butnot(Hammer.Ch_range('0', '9'), Hammer.Ch('6'));
              CheckParseOK(parser, "5", '5');
              CheckParseFail(parser, "6");
        }
        [Test]
        public void TestDifference() {
            Parser parser;
            parser = Hammer.Difference(Hammer.Token("ab"), Hammer.Ch('a'));
              CheckParseOK(parser, "ab", new byte[]{ 0x61, 0x62});
              CheckParseFail(parser, "a");
        }
        [Test]
        public void TestXor() {
            Parser parser;
            parser = Hammer.Xor(Hammer.Ch_range('0', '6'), Hammer.Ch_range('5', '9'));
              CheckParseOK(parser, "0", '0');
              CheckParseOK(parser, "9", '9');
              CheckParseFail(parser, "5");
              CheckParseFail(parser, "a");
        }
        [Test]
        public void TestMany() {
            Parser parser;
            parser = Hammer.Many(Hammer.Choice(Hammer.Ch('a'), Hammer.Ch('b')));
              CheckParseOK(parser, "", new object[]{ });
              CheckParseOK(parser, "a", new object[]{ 'a'});
              CheckParseOK(parser, "b", new object[]{ 'b'});
              CheckParseOK(parser, "aabbaba", new object[]{ 'a', 'a', 'b', 'b', 'a', 'b', 'a'});
        }
        [Test]
        public void TestMany1() {
            Parser parser;
            parser = Hammer.Many1(Hammer.Choice(Hammer.Ch('a'), Hammer.Ch('b')));
              CheckParseFail(parser, "");
              CheckParseOK(parser, "a", new object[]{ 'a'});
              CheckParseOK(parser, "b", new object[]{ 'b'});
              CheckParseOK(parser, "aabbaba", new object[]{ 'a', 'a', 'b', 'b', 'a', 'b', 'a'});
              CheckParseFail(parser, "daabbabadef");
        }
        [Test]
        public void TestRepeat_n() {
            Parser parser;
            parser = Hammer.Repeat_n(Hammer.Choice(Hammer.Ch('a'), Hammer.Ch('b')), 0x2);
              CheckParseFail(parser, "adef");
              CheckParseOK(parser, "abdef", new object[]{ 'a', 'b'});
              CheckParseFail(parser, "dabdef");
        }
        [Test]
        public void TestOptional() {
            Parser parser;
            parser = Hammer.Sequence(Hammer.Ch('a'), Hammer.Optional(Hammer.Choice(Hammer.Ch('b'), Hammer.Ch('c'))), Hammer.Ch('d'));
              CheckParseOK(parser, "abd", new object[]{ 'a', 'b', 'd'});
              CheckParseOK(parser, "acd", new object[]{ 'a', 'c', 'd'});
              CheckParseOK(parser, "ad", new object[]{ 'a', null, 'd'});
              CheckParseFail(parser, "aed");
              CheckParseFail(parser, "ab");
              CheckParseFail(parser, "ac");
        }
        [Test]
        public void TestIgnore() {
            Parser parser;
            parser = Hammer.Sequence(Hammer.Ch('a'), Hammer.Ignore(Hammer.Ch('b')), Hammer.Ch('c'));
              CheckParseOK(parser, "abc", new object[]{ 'a', 'c'});
              CheckParseFail(parser, "ac");
        }
        [Test]
        public void TestSepBy() {
            Parser parser;
            parser = Hammer.SepBy(Hammer.Choice(Hammer.Ch('1'), Hammer.Ch('2'), Hammer.Ch('3')), Hammer.Ch(','));
              CheckParseOK(parser, "1,2,3", new object[]{ '1', '2', '3'});
              CheckParseOK(parser, "1,3,2", new object[]{ '1', '3', '2'});
              CheckParseOK(parser, "1,3", new object[]{ '1', '3'});
              CheckParseOK(parser, "3", new object[]{ '3'});
              CheckParseOK(parser, "", new object[]{ });
        }
        [Test]
        public void TestSepBy1() {
            Parser parser;
            parser = Hammer.SepBy1(Hammer.Choice(Hammer.Ch('1'), Hammer.Ch('2'), Hammer.Ch('3')), Hammer.Ch(','));
              CheckParseOK(parser, "1,2,3", new object[]{ '1', '2', '3'});
              CheckParseOK(parser, "1,3,2", new object[]{ '1', '3', '2'});
              CheckParseOK(parser, "1,3", new object[]{ '1', '3'});
              CheckParseOK(parser, "3", new object[]{ '3'});
              CheckParseFail(parser, "");
        }
        [Test]
        public void TestAnd() {
            Parser parser;
            parser = Hammer.Sequence(Hammer.And(Hammer.Ch('0')), Hammer.Ch('0'));
              CheckParseOK(parser, "0", new object[]{ '0'});
              CheckParseFail(parser, "1");
            parser = Hammer.Sequence(Hammer.And(Hammer.Ch('0')), Hammer.Ch('1'));
              CheckParseFail(parser, "0");
              CheckParseFail(parser, "1");
            parser = Hammer.Sequence(Hammer.Ch('1'), Hammer.And(Hammer.Ch('2')));
              CheckParseOK(parser, "12", new object[]{ '1'});
              CheckParseFail(parser, "13");
        }
        [Test]
        public void TestNot() {
            Parser parser;
            parser = Hammer.Sequence(Hammer.Ch('a'), Hammer.Choice(Hammer.Token("+"), Hammer.Token("++")), Hammer.Ch('b'));
              CheckParseOK(parser, "a+b", new object[]{ 'a', new byte[]{ 0x2b}, 'b'});
              CheckParseFail(parser, "a++b");
            parser = Hammer.Sequence(Hammer.Ch('a'), Hammer.Choice(Hammer.Sequence(Hammer.Token("+"), Hammer.Not(Hammer.Ch('+'))), Hammer.Token("++")), Hammer.Ch('b'));
              CheckParseOK(parser, "a+b", new object[]{ 'a', new object[]{ new byte[]{ 0x2b}}, 'b'});
              CheckParseOK(parser, "a++b", new object[]{ 'a', new byte[]{ 0x2b, 0x2b}, 'b'});
        }
        [Test]
        public void TestRightrec() {
            Parser parser;
            IndirectParser sp_rr = Hammer.Indirect();
            sp_rr.Bind(Hammer.Choice(Hammer.Sequence(Hammer.Ch('a'), sp_rr), Hammer.Epsilon_p()));
            parser = sp_rr;
              CheckParseOK(parser, "a", new object[]{ 'a'});
              CheckParseOK(parser, "aa", new object[]{ 'a', new object[]{ 'a'}});
              CheckParseOK(parser, "aaa", new object[]{ 'a', new object[]{ 'a', new object[]{ 'a'}}});
        }
        [Test]
        public void TestAmbiguous() {
            Parser parser;
            IndirectParser sp_d = Hammer.Indirect();
            IndirectParser sp_p = Hammer.Indirect();
            IndirectParser sp_e = Hammer.Indirect();
            sp_d.Bind(Hammer.Ch('d'));
            sp_p.Bind(Hammer.Ch('+'));
            sp_e.Bind(Hammer.Choice(Hammer.Sequence(sp_e, sp_p, sp_e), sp_d));
            parser = sp_e;
              CheckParseOK(parser, "d", 'd');
              CheckParseOK(parser, "d+d", new object[]{ 'd', '+', 'd'});
              CheckParseOK(parser, "d+d+d", new object[]{ new object[]{ 'd', '+', 'd'}, '+', 'd'});
        }
    }
}
